SCF/OOP_Attempt/Epack.py

```diff
@@ -0,0 +1,284 @@
+        for i in range(len(Cn) - 1):
+            F += Cn[i] * trials[i]
+
+        return F
+        #Diagonalize F and Reproduce Density Matrix after!!
+    
+    def SOSCF(self, C, F, Qpq, ndocc, nvirt):
+        # Remember only the OV part is needed.
+        grad = -4 * np.dot(C.T, F).dot(C)[:ndocc, ndocc:]
+        Fmo = np.einsum('ij,jk,kl->il', C.T, F, C)              
+                   
+        eps = np.diag(Fmo)
+        precon = -4 * (eps[:ndocc][:, None] - eps[ndocc:])
+        
+        k = -grad / precon
+        p = []
+        rk = []
+        z = []
+        
+        Cocc = C[:,:ndocc] 
+        Cvirt = C[:,ndocc:]
+        Ax = self.build_Ax(Fmo, Cocc, k, Cvirt, Qpq).ravel() #Build DF Hessian
+        
+        k = k.ravel()
+        rk.append(-grad.ravel() - Ax)
+        z.append(rk[0]/precon.ravel())
+        p.append(z[0])
+
+        grad_rms = np.mean( grad ** 2) ** 0.5 #Relative metric
+        microstart = time.time()
+
+        for j in range(self.maxmicro):
+
+            tmp = self.build_Ax(Fmo, Cocc, p[j].reshape(ndocc, nvirt), Cvirt, Qpq).ravel()                   
+
+            alpha = np.dot(rk[j].T, z[j])
+            alpha /= np.dot(p[j].T, tmp)
+             
+            k = k + alpha*p[j]
+            rk.append(rk[j] - alpha*tmp)
+             
+            err = np.mean(rk[j+1]**2)**0.5
+            err = err / grad_rms
+            print ("SOSCF Microiteration %d: RMS = %2.5E" % (j+1, err))
+            self.totmicro += 1 
+            if err < self.SOSCF_tol or j == self.maxmicro - 1:
+                
+                if j == self.maxmicro - 1:
+                    print ("SOSCF could not converge and was turned OFF\n")
+                    soscf = False
+                else:
+                    soscf = True
+                break
+
+            z.append(rk[j+1]/precon.ravel())
+            beta = np.dot(z[j+1].T, rk[j+1]) / np.dot(z[j].T,rk[j])
+            p.append(z[j+1] + beta*p[j])
+
+        self.microtime += (time.time() - microstart)
+        # Convert back, mind the compound indices
+        k.shape = (ndocc, nvirt)
+
+        # Build the full Kappa matrix
+        full_K = np.zeros((self.nbf, self.nbf))
+        full_K[:ndocc, ndocc:] = k
+        full_K[ndocc:, :ndocc] = -k.T
+        # Construct U and new orbitals
+        U = sl.expm(full_K)
+        C = np.dot(C, U)
+        # Update other matrices
+        Cocc = C[:, :ndocc].copy()
+        Cvirt = C[:, ndocc:].copy()
+        D = np.dot(Cocc, Cocc.T)
+
+        return C, Cocc, Cvirt, D, soscf
+
+    def build_Ax(self, Fmo, Cocc, kappa, Cvirt, Qpq):
+       
+        ndocc = np.shape(Cocc)
+        Ax = np.dot(Fmo[:ndocc[1], :ndocc[1]], -kappa)
+        Ax += np.dot(Fmo[ndocc[1]:, ndocc[1]:], kappa.T).T
+        Cright = np.dot(kappa, Cvirt.T).T
+        J, K = self.JK.build_DFJK(Qpq, Cocc, Cright)
+        Ax += np.dot(Cocc.T, (4*J - K - K.T)).dot(Cvirt)
+
+        return Ax*4
+
+
+    def energy(self, etype):
```

Comment:
I would rename this to something like `compute_energy` or maybe `iterate`. `energy` seems more like an access to the energy of the object.